pathdb/datastax/src/main/java/org/commonjava/storage/pathmapped/pathdb/datastax/CassandraPathDB.java

```diff
@@ -154,6 +164,60 @@ public Session getSession()
         return session;
     }
 
+    @Override
+    public Set<String> getFileSystemContaining( Collection<String> candidates, String path )
+    {
+        logger.debug( "Get fileSystem containing path {}, candidates: {}", path, candidates );
+        if ( ROOT_DIR.equals( path ) )
+        {
+            return emptySet();
+        }
+        String parentPath = PathMapUtils.getParentPath( path );
+        String filename = PathMapUtils.getFilename( path );
+
+        BoundStatement bound = preparedContainingQuery.bind( candidates, parentPath, filename );
+        ResultSet result = session.execute( bound );
+        return result.all().stream().map( row -> row.get( 0, String.class ) ).collect( Collectors.toSet() );
+    }
+
+    @Override
+    public Set<String> getFileSystemContainingDirectory( Collection<String> candidates, String path )
+    {
+        if ( !path.endsWith( "/" ) )
+        {
+            path += "/";
+        }
+        return getFileSystemContaining( candidates, path );
+    }
+
+    /**
+     * Get the first fileSystem in the candidates containing the path.
+     *
+     * The CQL query results are not returned in the order in which the key was specified in the IN clause.
+     * The results are returned in the natural order of the column so we can not rely on the query like '...limit 1'.
+     */
+    @Override
+    public String getFirstFileSystemContaining( List<String> candidates, String path )
+    {
+        logger.debug( "Get first fileSystem containing path {}, candidates: {}", path, candidates );
+        List<List<String>> batches = Lists.partition( candidates, DEFAULT_GET_FIRST_BATCH_SIZE );
+        for ( List<String> batch : batches )
+        {
+            Set<String> ret = getFileSystemContaining( batch, path );
+            if ( !ret.isEmpty() )
+            {
+                for ( String candidate : batch )
+                {
+                    if ( ret.contains( candidate ) )
+                    {
+                        return candidate;
```

Comment:
So the most bad edge case here will cause a loop of the cadidates.size() times looping I think. But I'm not not sure how to optimize here if we really want to get the first by natural order from candidates too. Maybe we can leave a comment here for future optimization.

Reply:
I once thought the datastax run the query by the order in IN clause so that the "select ... limit 1" will have better performance. That is why this api came. 
But it is not. Cassandra ignore the IN order. So this API does not make much sense. 
I use a 1000 batch size. For cases when candidates are less than 1000, this can be used as a convinient method. For cases we have 3000 repos, the api helps if the targe is in the first 1000 or 2000 repos. If it is not exist or close to the end of array, this api is helpless and a little slower than the normal api. But difference is little. I pushed a new test report and some explanation there in the QueryByPathPerformanceTest.

Reply:
But if it is confusing, I am Ok to remove the batching thing. Only use it as a convinent method.

Reply:
I remove the batching thing. It only complicate the code.